File: android/hakka-network/src/main/kotlin/com/noodleapps/hakka/PluginRegistry.kt
package com.noodleapps.hakka

import java.util.concurrent.CopyOnWriteArrayList

/**
 * Manages [HakkaPlugin] registration for a single [HakkaInterceptor] instance.
 *
 * Mirrors `Hakka.use(plugin)` from the TypeScript core (packages/hakka-core/src/engine/plugins.ts):
 *  - registration is idempotent per plugin id
 *  - [setup] is called exactly once per plugin with a live [HakkaPluginContext]
 *  - teardown lambdas returned by [setup] are stored and called on [removeAll]
 *
 * Obtain via [HakkaInterceptor.plugins].
 */
class PluginRegistry internal constructor(
    private val contextProvider: () -> HakkaPluginContext,
) {
    /**
     * Ordered list of registered entries. CopyOnWriteArrayList for safe iteration while
     * registrations happen from any thread (registrations are rare; reads are common).
     */
    private val entries = CopyOnWriteArrayList<Entry>()

    private data class Entry(
        val plugin: HakkaPlugin,
        val teardown: (() -> Unit)?,
    )

    /**
     * Register [plugin] with the engine. Idempotent — if a plugin with the same [HakkaPlugin.id]
     * is already registered, this call is a no-op and returns false.
     *
     * [HakkaPlugin.setup] is called synchronously on the calling thread.
     *
     * @return true if the plugin was registered, false if it was already present.
     */
    fun use(plugin: HakkaPlugin): Boolean {
        if (entries.any { it.plugin.id == plugin.id }) return false
        val ctx = contextProvider()
        val teardown = try {
            plugin.setup(ctx)
        } catch (_: Exception) {
            null
        }
        entries.add(Entry(plugin, teardown))
        return true
    }

    /**
     * Remove the plugin with [pluginId] and invoke its teardown (if any).
     * Returns true if the plugin was found and removed.
     */
    fun remove(pluginId: String): Boolean {
        val entry = entries.firstOrNull { it.plugin.id == pluginId } ?: return false
        entries.remove(entry)
        try { entry.teardown?.invoke() } catch (_: Exception) {}
        return true
    }

    /** All panels contributed by registered plugins, sorted ascending by [HakkaPlugin.id] stability
     *  (stable order within a session; the host is responsible for final ordering). */
    fun registeredPlugins(): List<HakkaPlugin> = entries.map { it.plugin }

    /** Run all teardowns and clear the registry. Called when the interceptor is closed. */
    internal fun removeAll() {
        val snapshot = entries.toList()
        entries.clear()
        snapshot.forEach { e ->
            try { e.teardown?.invoke() } catch (_: Exception) {}
        }
    }
}

File: ios/Sources/Common/PluginRegistry.swift
import Foundation

// MARK: - HakkaPluginRegistry

/// Thread-safe store of registered plugins. Owned by ``HakkaInterceptor``.
///
/// Registration is idempotent by plugin `id` — a second `use(_:)` call with
/// the same id is silently ignored (mirrors the TS engine behaviour).
public final class HakkaPluginRegistry: @unchecked Sendable {
    private let lock = NSLock()
    private var entries: [String: Entry] = [:]

    /// Ordered list of all registered plugins (registration order preserved).
    public var plugins: [any HakkaPlugin] {
        lock.lock()
        defer { lock.unlock() }
        return entries.values.sorted { $0.index < $1.index }.map(\.plugin)
    }

    /// All panels contributed by registered plugins, sorted by `order` then
    /// registration order (stable sort).
    public var panels: [HakkaPanel] {
        lock.lock()
        defer { lock.unlock() }
        return entries.values
            .sorted { $0.index < $1.index }
            .flatMap { $0.plugin.panels }
            .sorted { $0.order < $1.order }
    }

    public init() {}

    // MARK: - Registration

    /// Register `plugin`, calling `setup(ctx:)` immediately with the provided
    /// context. No-op if a plugin with the same `id` is already registered.
    ///
    /// Returns `true` when the plugin was newly registered, `false` on duplicate.
    @discardableResult
    public func use(_ plugin: any HakkaPlugin, ctx: any HakkaPluginContext) -> Bool {
        lock.lock()
        guard entries[plugin.id] == nil else {
            lock.unlock()
            return false
        }
        let index = entries.count
        entries[plugin.id] = Entry(plugin: plugin, index: index)
        lock.unlock()

        plugin.setup(ctx: ctx)
        return true
    }

    // MARK: - Private

    private struct Entry {
        let plugin: any HakkaPlugin
        let index: Int
    }
}
